test/server_test.lua

```diff
@@ -208,10 +208,11 @@ g.test_net_box_exec = function()
 
     do
         local _l_fn, fn = l(), function() error('X_x') end
-        t.assert_error_msg_equals(
-            efmt(_l_fn, 'X_x'),
-            exec, fn
-        )
```

Comment:
It looks dangerous. Shouldn't this helper understands luatest's wrapping and unwrap it automatically?

The same regarding `t.assert_error_msg_contains` and other similar error capturing methods.

I guess that this behavior change may hit many tests.

Reply:
> \- Make `assert_error_*` ignore errors raised by assertion failures.

I would say 're-raise' instead of 'ignore' here to make the new behavior more obvious from the text.

----

The idea to re-raise assertions is interesting.

It seems safe in context of `t.assert_error_msg_*{equals,contains,matches}`, because AFAIS a luatest error table has no `__tostring` metamethod and these functions never accept these errors before.

The things are different for `t.assert_error` (and maybe for `t.assert_error_covers`). For example, `test_bs_leader_not_goes_ro_if_appointed` from https://github.com/tarantool/tarantool-ee/pull/863 has a retrying helper, which uses `t.assert_equals` in its predicate (to get a readable errors in case of a failure), and the test case verifies that the retrying fails using `t.assert_error`.

Also, sometimes I use a plain `assert` to verify test case preconditions and use `t.assert_*` to verify what the test case aims to check. In this situation the error from the plain assert is more severe.

I'm in some doubt here.

I propose to discuss other possible solutions. What comes into my head:

* Tune `t.assert_error_msg_*{equals,contains,matches}` to look into the `message` field of the error if it is a table.
* Rename a general-looking `{class = 'LuatestError', status = 'error', <...>}` to something like `{class = 'LuatestError', status = 'wrapped_by_exec', <...>}` and unwrap it in `t.assert_error_*` before any comparisons.

Or maybe I need to spin it in the head a bit more. For a while, just wrote down my current thoughts.

Reply:
> The things are different for `t.assert_error` (and maybe for `t.assert_error_covers`). For example, `test_bs_leader_not_goes_ro_if_appointed` from [tarantool/tarantool-ee#863](https://github.com/tarantool/tarantool-ee/pull/863) has a retrying helper, which uses `t.assert_equals` in its predicate (to get a readable errors in case of a failure), and the test case verifies that the retrying fails using `t.assert_error`.

If you meant this

```lua
    -- In case of long yield the test may finish as a
    -- false positive. Though it's very unlikely.
    t.assert_error(wait_database_mode, g, 'instance-001', 'ro')
```

I would rewrite that test because as it is now it would pass even if the tested code throws a different kind of error (e.g. a Lua assertion or syntax error). IMO we should raise a proper error in this test and check that the expected error is raised. Actually, I'd deprecate `t.assert_error` altogether for this reason - it's too dangerous. I myself never use it and most Tarantool developers don't, either:

```
# Ran in Tarantool repo:
$ grep -r '\<assert_error\>' test | wc -l
15
```

> Also, sometimes I use a plain `assert` to verify test case preconditions and use `t.assert_*` to verify what the test case aims to check. In this situation the error from the plain assert is more severe.

Why? Luatest assertions are much more powerful and result in much more verbose error messages.

> I propose to discuss other possible solutions. What comes into my head:
> 
> * Tune `t.assert_error_msg_*{equals,contains,matches}` to look into the `message` field of the error if it is a table.
> * Rename a general-looking `{class = 'LuatestError', status = 'error', <...>}` to something like `{class = 'LuatestError', status = 'wrapped_by_exec', <...>}` and unwrap it in `t.assert_error_*` before any comparisons.

We could do that or we could just rewrite `pcall_wrapper`introduced in this patch  to only unwrap `status = 'error'` errors (they aren't normally raised by luatest), without re-raising failures. However, I prefer to make `t.assert_error_*` let assertion failures go through because:
- Catching assertion failures and checking their message in tests makes no sense.
- Catching assertion failures with `t.assert_error` is dangerous because it catches all kinds of errors.
- If we let `t.assert_error_*` catch an assertion failure (as it does now), we'd get a wrong error and trace: instead of the original assertion failure, we'd see something like "Error message expected: ... Error message received: expected: foo actual: bar", which is confusing.

Reply:
Sounds meaningful, thanks for the clarification.

I'm OK, but, please, write the new rules of the game somewhere: in README, for example. It shouldn't be an implicit knowledge.

> Actually, I'd deprecate `t.assert_error` altogether for this reason - it's too dangerous.

It feels valid for me too. At least, I would write a recommendation to wait certain errors and discourage `t.assert_error` or a catching-all `pcall` use. Fits good to README as well, IMHO.

----

> I would rewrite that test because as it is now it would pass even if the tested code throws a different kind of error (e.g. a Lua assertion or syntax error). IMO we should raise a proper error in this test and check that the expected error is raised.

I agree. `t.assert_equals` is attractive, because I don't need to bother with constructing a good error message. OTOH, it seems good to write a message that is more specific for this retrying helper.

@georgiy-belyanin Please, take a look on our discussion.

----

Disregarding of the main discussion topic `status = 'error'` feels too general and we can name it in a somehow more specificaly to the exec's wrapping. Or clarify somewhere in the code, what does `status = 'error'` means and when it should/shouldn't be used.

I guess you have an idea about some `assert`/`test status`/`error` separation, but it seems quite implicit now. At least, I'm not familiar with luatest's code enough.

Reply:
> * If we let `t.assert_error_*` catch an assertion failure (as it does now), we'd get a wrong error and trace: instead of the original assertion failure, we'd see something like "Error message expected: ... Error message received: expected: foo actual: bar", which is confusing.

Come to think of it, my patch actually does overwrite the assertion failure trace by re-raising the error :(

So I think I'll go with this proposition of your for now and won't re-raise assertion failure:

> * Rename a general-looking `{class = 'LuatestError', status = 'error', <...>}` to something like `{class = 'LuatestError', status = 'wrapped_by_exec', <...>}` and unwrap it in `t.assert_error_*` before any comparisons.

Reply:
> Disregarding of the main discussion topic `status = 'error'` feels too general and we can name it in a somehow more specificaly to the exec's wrapping. Or clarify somewhere in the code, what does `status = 'error'` means and when it should/shouldn't be used.
> 
> I guess you have an idea about some `assert`/`test status`/`error` separation, but it seems quite implicit now. At least, I'm not familiar with luatest's code enough.

Agree, I introduced a new error class for errors wrapped by `Server:exec()`. Don't know why I decided to reuse luatest status in the first place - it does look confusing.